File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.py
from collections import deque
def solution(maps):
    dx = [-1,0,1,0]
    dy = [0,1,0,-1]

    r = len(maps[0])
    c = len(maps)
    
    def BFS(i,j,word):
        q = deque()
        q.append((i,j,0))
        while q:
            x, y, cnt = q.popleft()
            if maps[x][y] == word:
                if word == 'L':
                    return (x,y,cnt)
                elif word =='E':
                    return x,y,cnt
            for k in range(4):
                nx = x + dx[k]
                ny = y + dy[k]
                if 0<=nx<c and 0<=ny<r:
                    if not visited[nx][ny] and maps[nx][ny]!='X':
                        visited[nx][ny] = True
                        q.append((nx,ny,cnt+1))
        return x,y, False
    
    visited = [[False] * r for _ in range(c)]
    for i in range(c):
        for j in range(r):
            if maps[i][j]=='S':
                x,y,L= BFS(i,j,'L')
                if not L:
                    return -1
                else:
                    visited = [[False] * r for _ in range(c)]
                    x,y,E = BFS(x,y,'E')
                    if not E:
                        return -1
                    else:
                        return E+L
                    

File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.js
function solution(maps) {
    const dir = [[0,1],[0,-1],[1,0],[-1,0]]
    const r = maps.length
    const c = maps[0].length
    const start = [0,0]
    const lever = [0,0]
    const end = [0,0] 
    
    for(let i = 0; i < r; i++){
        for(let j = 0; j < c; j++){
            if(maps[i][j] === "S"){
                start[0] = i
                start[1] = j 
            }
            if(maps[i][j] === "L"){
                lever[0] = i
                lever[1] = j 
            }
            if(maps[i][j] === "E"){
                end[0] = i
                end[1] = j 
            }
        }
    }
    const distanceToLever = BFS(start[1], start[0], lever[1], lever[0])
    const distanceToEnd = BFS(lever[1], lever[0], end[1], end[0])
    
    return distanceToLever=== -1 || distanceToEnd === -1 ? -1 : distanceToLever + distanceToEnd;

    
    function BFS(sX, sY, eX, eY){
        const visited = Array.from({length:r}, () => Array(c).fill(false))
        const queue = []
        queue.push([sY, sX, 0])
        visited[sY][sX] = true
        let begin = 0
        while(begin < queue.length){
            // 현재 위치, 이동 횟수
            const [cY, cX, cnt] = queue[begin++]
            
            // 목적지 도착
            if(cY===eY && cX ===eX) return cnt
            
            // 4방향으로 이동
            for(let k = 0; k < 4; k++){
                const ny = cY + dir[k][0]
                const nx = cX + dir[k][1]
                if(ny < 0 || ny >= r || nx < 0 || nx >=c || maps[ny][nx]==="X" || visited[ny][nx]) continue
                visited[ny][nx] = true
                queue.push([ny, nx, cnt + 1])
            }
        }
        
        return -1
    }
}


